src/Xamarin.Android.Tools.Bytecode/JavaParameterNamesLoader.cs

```diff
@@ -0,0 +1,192 @@
+
+			var methods = api.XPathSelectElements ("package/*/*");
+			foreach (var method in methods) {
+				switch (method.Name.LocalName) {
+				case "method":
+				case "constructor":
+
+					string package = method.Parent.Parent.Attribute ("name").Value;
+					string type = method.Parent.Attribute ("name").Value;
+					string mname = method.Attribute ("name").Value;
+					var parameters = method.Elements ("parameter").ToArray ();
+					if (!parameters.Any ()) // we don't care about parameterless methods.
+						continue;
+					var matchedPackage = fixup.FirstOrDefault (p => p.Name == package);
+					if (matchedPackage == null) {
+						Log.Warning (0, "Package {0} not found.", package);
+						continue;
+					}
+					var matchedType = matchedPackage.Types.FirstOrDefault (t => t.Name == type);
+					if (matchedType == null) {
+						Log.Warning (0, "Type {0} not found.", package + '.' + type);
+						continue;
+					}
+					var matchedMethods = matchedType.Methods.Where (m => (m.Name == "#ctor" ? method.Name.LocalName == "constructor" : m.Name == mname) && m.Parameters.Count == parameters.Length);
+					if (!matchedMethods.Any ()) {
+						Log.Warning (0, "Method {0} with {1} parameters not found.", package + '.' + type + '.' + mname, parameters.Length);
+						continue;
+					}
+					var matched = matchedMethods.FirstOrDefault (m => m.Parameters.Zip (parameters, (f, x) => f.Type == x.Attribute ("type").Value.Replace (", ", ",")).All (b => b));
+
+					if (matched == null) {
+						Log.Warning (0, "Method {0}({1}) not found.",
+						             package + '.' + type + '.' + mname,
+						             string.Join (",", parameters.Select (para => para.Attribute ("type").Value)));
+						continue;
+					}
+
+					for (int i = 0; i < parameters.Length; i++)
+						parameters [i].SetAttributeValue ("name", matched.Parameters [i].Name);
+
+					matched.Applied = true;
+					break;
+				}
+			}
+			foreach (var p in fixup)
+				foreach (var t in p.Types)
+					foreach (var m in t.Methods.Where (m => !m.Applied))
+						Log.Warning (0, "Method parameter description for {0}.{1}.{2}({3}) is never applied",
+						             p.Name, t.Name, m.Name, string.Join (",", m.Parameters.Select (para => para.Type)));
+		}
+
+		class Parameter
+		{
+			public string Type { get; set; }
+			public string Name { get; set; }
+		}
+
+		class Method
+		{
+			public string Name { get; set; }
+			public List<Parameter> Parameters { get; set; }
+			public bool Applied { get; set; }
+		}
+
+		class Type
+		{
+			public string Name { get; set; }
+			public List<Method> Methods { get; set; }
+		}
+
+		class Package
+		{
+			public string Name { get; set; }
+			public List<Type> Types { get; set; }
+		}
+
+		// from https://github.com/atsushieno/xamarin-android-docimporter-ng/blob/83ed572ebfdcec8a8becd3337943a488bf56d57d/Xamarin.Android.Tools.JavaStubImporter/JavaApiParameterNamesXmlExporter.cs#L78
+		/*
+		 * The Text Format is:
```

Comment:
Did this format come from anywhere? Any particular reason to prefer a "baroque" plain-text format and not XML, or JSON?

Reply:
It's less than half size as compared to XML, and this format is 100x more readable, nicer for text searching, than XML or JSON.

There is actually XML output support in the description generator tool (and even consumer in this patch). But unless there is any reason there is no need to switch to less readable, less searchable formats, we should prefer developer-friendly text.

Google also uses similar text format to list the stable API in android_framework_base.

Reply:
> Google also uses similar text format to list the stable API in android_framework_base.

Link? :-)

Are there any differences? Could/should this format be made closer/identical?

Reply:
This: https://raw.githubusercontent.com/aosp-mirror/platform_frameworks_base/master/api/current.txt (remember there used to be api.xml which you guys basically took for generator.exe back in 2010?)

The format is of course not the same as the entire purpose is different.